### DISCUSSION TITLE
How does real-time update detection and frontend-backend integration work?

### DISCUSSION THREAD

**Discussion by patelka2211** [gitbutlerapp/gitbutler#4566] [Q&A] [answered]
Hi everyone, I’m curious about how GitButler detects updates in the repository in real time. For instance, if I change the content of a file in the repository, how does GitButler reflect this change almost instantly? Additionally, could someone explain how the backend integrates with the frontend to enable this real-time update? Any detailed insights would be really appreciated. Thanks!

**Comment by Byron** [comment, marked as answer]
It's always good to be curious, thanks for asking!

The short version is that [here is the entrypoint](https://github.com/gitbutlerapp/gitbutler/blob/2d82b6e038876d722e0ef8f7b0fc79843da1079b/crates/gitbutler-watcher/src/lib.rs#L74), which starts a file-monitor in the background which receives filesystem events. These events are then processed and posted to the frontend which responds to them.

That's already it, and you would be able to follow along by starting at the entrypoint above.

**Reply by patelka2211** [reply]
Thank you @Byron for the explanation!

As a beginner in Rust, this seems a bit overwhelming at the moment, but I’m excited to learn more and get up to speed. I’m looking forward to contributing to the project once I have a better understanding. I also learn by watching your videos from "Learn Rust with Gitoxide" playlist.

If you have anything more to share, please do. If not, we can close this thread.

Thanks again :) <3

**Reply by Byron** [reply]
Thanks for sharing! It would be great to see you contribute when you feel comfortable, and of course, that you are watching these videos and find them useful!

One more recommendation certainly would be to 'read, modify and run'. Reading code is great, but how can you predict what it does if it's not being run, maybe with modifications? Running with a debugger attached might also help as it allows introspection, even though I personally am either writing tests or using `dbg!(value)`.

Generally, the best way to learn for me is to try to make modifications, maybe it's something that can also work for you.
